File: 47-permutations-ii/47-permutations-ii.ts
function permuteUnique(nums: number[]): number[][] {
    if (nums.length === 1) {
        return [nums]
    }

    let res: number[][] = []
    let prev: number | null = null;
    nums.sort()
    for (let i = 0; i < nums.length; i++) {
        const p = nums[i]
        if (p == prev) continue;
        prev = p;
        const restNums = [...nums]
        restNums.splice(i, 1)
        const perms = permuteUnique(restNums).map(x => [p, ...x])
        res.push(...perms)
    }
    return res
}

File: 47-permutations-ii/47-permutations-ii.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        set <vector <int>> perms;
        do{
            perms.insert(nums);    
        }while( next_permutation(nums.begin(),nums.end()));
        return vector <vector <int>> (perms.begin(), perms.end());
    }
};
